contrib/babelfishpg_common/src/varbinary.c

```diff
@@ -697,21 +718,45 @@ Datum
 varbinaryvarchar(PG_FUNCTION_ARGS)
 {
 	bytea	   *source = PG_GETARG_BYTEA_PP(0);
-	char	   *data = VARDATA_ANY(source);
+	char	   *data = VARDATA_ANY(source);		/* Source data is server encoded */
+	VarChar    *result;
+	char 	   *encoded_result;
 	size_t		len = VARSIZE_ANY_EXHDR(source);
 	int32		typmod = PG_GETARG_INT32(1);
 	int32		maxlen = typmod - VARHDRSZ;
-	VarChar    *result;
+	coll_info	collInfo;
+	int			encodedByteLen;
 
 	/*
-	 * Cast the entire input binary data if maxlen is invalid or supplied data
-	 * fits it
+	 * Allow trailing null bytes 
+	 * Its safe since multi byte UTF-8 does not contain 0x00 
 	 */
-	if (maxlen < 0 || len <= maxlen)
-		result = (VarChar *) cstring_to_text_with_len(data, len);
-	/* Else truncate it */
-	else
-		result = (VarChar *) cstring_to_text_with_len(data, maxlen);
+	while(len>0 && data[len-1] == '\0')
+		len -= 1;
+	
+	collInfo = lookup_collation_table(get_server_collation_oid_internal(false));
+	/*
+	 * Cast the entire input binary data if maxlen is 
+	 * invalid or supplied data fits it
+	 * Else truncate it
+	 */
+	PG_TRY();
+	{
+		if (maxlen < 0 || len <= maxlen)
+			encoded_result = encoding_conv_util(data, len, collInfo.enc, PG_UTF8, &encodedByteLen);
+		else
+			encoded_result = encoding_conv_util(data, maxlen, collInfo.enc, PG_UTF8, &encodedByteLen);
+	}
+	PG_CATCH();
+	{
+		ereport(ERROR,
+			   (errcode(ERRCODE_INTERNAL_ERROR),
+				errmsg("Failed to convert from data type varbinary to varchar")));
```

Comment:
we somehow need to preserve the original error as log or error in the logfile

Reply:
throwing the original error alongside are our own.